cpp/tests/io/arrow_io_source_test.cpp

```diff
@@ -87,7 +87,18 @@ TEST_F(ArrowIOTest, S3FileSystem)
     ASSERT_EQ(1, tbl.tbl->num_columns());  // Only single column specified in reader_options
     ASSERT_EQ(244, tbl.tbl->num_rows());   // known number of rows from the S3 file
   }
-  CUDF_EXPECTS(arrow::fs::EnsureS3Finalized().ok(), "Failed to finalize s3 filesystem");
+  if (!s3_unsupported) {
+    // Verify that we are using Arrow with S3, and call finalize
+    // https://github.com/apache/arrow/issues/36974
+    // needs to be in a separate conditional to ensure we call
+    // finalize after all arrow_io_source's has been deleted
+    void* whole_app                                       = dlopen(NULL, RTLD_LAZY);
+    decltype(arrow::fs::EnsureS3Finalized)* close_s3_func = nullptr;
+
+    close_s3_func = reinterpret_cast<decltype(close_s3_func)>(
+      dlsym(whole_app, "_ZN5arrow2fs17EnsureS3FinalizedEv"));
+    if (close_s3_func) { CUDF_EXPECTS(close_s3_func().ok(), "Failed to finalize s3 filesystem"); }
+  }
```

Comment:
Should this block call `dlclose(whole_app);` for completeness?

Reply:
Yes it should. Done